### PR TITLE
[Support inference params-1]Add ParamSchema and ParamSpec in model signature

### PR DESCRIPTION
**Pull request by serena-ruan**
## Related Issues/PRs

Split 1: #8563

## What changes are proposed in this pull request?
Add ParamSchema in ModelSignature

## How is this patch tested?

- [x] Existing unit/integration tests
- [x] New unit/integration tests
- [ ] Manual tests (describe details, including test results, below)

## Does this PR change the documentation?

- [x] No. You can skip the rest of this section.
- [ ] Yes. Make sure the changed pages / sections render correctly in the documentation preview.

### Is this a user-facing change?

- [ ] No. You can skip the rest of this section.
- [x] Yes. Give a description of this change to be included in the release notes for MLflow users.

Add `params` field in ModelSignature, for further supports of `params` in inference.

### What component(s), interfaces, languages, and integrations does this PR affect?

Components

- [ ] `area/artifacts`: Artifact stores and artifact logging
- [ ] `area/build`: Build and test infrastructure for MLflow
- [ ] `area/docs`: MLflow documentation pages
- [ ] `area/examples`: Example code
- [ ] `area/model-registry`: Model Registry service, APIs, and the fluent client calls for Model Registry
- [x] `area/models`: MLmodel format, model serialization/deserialization, flavors
- [ ] `area/recipes`: Recipes, Recipe APIs, Recipe configs, Recipe Templates
- [ ] `area/projects`: MLproject format, project running backends
- [ ] `area/scoring`: MLflow Model server, model deployment tools, Spark UDFs
- [ ] `area/server-infra`: MLflow Tracking server backend
- [ ] `area/tracking`: Tracking Service, tracking client APIs, autologging

Interface

- [ ] `area/uiux`: Front-end, user experience, plotting, JavaScript, JavaScript dev server
- [ ] `area/docker`: Docker use across MLflow's components, such as MLflow Projects and MLflow Models
- [ ] `area/sqlalchemy`: Use of SQLAlchemy in the Tracking Service or Model Registry
- [ ] `area/windows`: Windows support

Language

- [ ] `language/r`: R APIs and clients
- [ ] `language/java`: Java APIs and clients
- [ ] `language/new`: Proposals for new client languages

Integrations

- [ ] `integrations/azure`: Azure and Azure ML integrations
- [ ] `integrations/sagemaker`: SageMaker integrations
- [ ] `integrations/databricks`: Databricks integrations

<a name="release-note-category"></a>

### How should the PR be classified in the release notes? Choose one:

- [ ] `rn/breaking-change` - The PR will be mentioned in the "Breaking Changes" section
- [ ] `rn/none` - No description will be included. The PR will be mentioned only by the PR number in the "Small Bugfixes and Documentation Updates" section
- [x] `rn/feature` - A new user-facing feature worth mentioning in the release notes
- [ ] `rn/bug-fix` - A user-facing bug fix worth mentioning in the release notes
- [ ] `rn/documentation` - A user-facing documentation change worth mentioning in the release notes

### CODE REVIEW

File: mlflow/types/schema.py
Comment (serena-ruan) [inline mlflow/types/schema.py:539]: Should we always convert type to numpy type here? For example python int can map to int32 or int64, if we don't convert then ParamSpec("a", DataType.long, 1) and ParamSpec("a", DataType.integer, 1) both work, while if we convert then they will become ParamSpec("a", DataType.long, np.int64(1)) and ParamSpec("a", DataType.integer, np.int32(1)).

**Comment by serena-ruan** [comment]
This is the first split of that giant PR, and I created 5 which the later depends on the former. I could either 1. wait until the former PR is merged then clean up the next PR; or 2. make the next PR against the former PR's branch. WDYT?

File: mlflow/types/schema.py
Comment (harupy) [reply mlflow/types/schema.py:539]: For scalar types, using python native type seems safer.

Comment (harupy) [reply mlflow/types/schema.py:539]:
```python
>>> isinstance(np.int64(1), int)
False
>>> isinstance(np.int32(1), int)
False
>>> isinstance(1, int)
True
```

File: mlflow/models/signature.py
Comment (BenWilson2) [inline mlflow/models/signature.py:73]: silly nit:

```suggestion
                "if params are provided, they must be of type mlflow.models.signature.ParamSchema. "
                "got '{}'".format(type(params))
```
I know this mirrors the TypeError from above, but that one also is a little confusing to a user :)

File: mlflow/models/utils.py
Comment (harupy) [inline mlflow/models/utils.py:944]: /ai give feedback on this function

File: tests/pyfunc/test_pyfunc_schema_enforcement.py
Comment (harupy) [inline tests/pyfunc/test_pyfunc_schema_enforcement.py:1070]: /ai Can you split this test into smaller tests?

File: mlflow/types/schema.py
Comment (BenWilson2) [inline mlflow/types/schema.py:472]: Could we use a different name here? `param_type`? Overriding a reserved func can be a bit dangerous / hard to follow.

Comment (harupy) [reply mlflow/types/schema.py:472]: +1

File: mlflow/types/schema.py
Comment (BenWilson2) [inline mlflow/types/schema.py:616]: I think it would be a little less confusing to name this something other than the shadow of the builtin

File: mlflow/types/schema.py
Comment (harupy) [reply mlflow/types/schema.py:472]: or `date_type`?

Comment (harupy) [reply mlflow/types/schema.py:472]: or `dtype`?

File: mlflow/types/schema.py
Comment (BenWilson2) [inline mlflow/types/schema.py:531]: do we need to state `value` here?

```
if value is None:
    return
```

File: mlflow/types/schema.py
Comment (BenWilson2) [inline mlflow/types/schema.py:536]:
```suggestion
            raise MlflowException.invalid_parameter_value(
                f"Value should be a scalar for param {name}, got {value}"
            )
```
In case you don't want to import that constant :)

File: mlflow/types/schema.py
Comment (BenWilson2) [inline mlflow/types/schema.py]: Does this work for all precisions? Just making sure that there isn't some lossy inference operation that happens with the changes that were made to numpy last year with datetime value processing

File: mlflow/types/schema.py
Comment (BenWilson2) [inline mlflow/types/schema.py:578]: do we need to use Tuple here or would List suffice? I'm wary of non-JSON-serializable types if ever the shape spec would need to be encoded

File: mlflow/types/schema.py
Comment (BenWilson2) [inline mlflow/types/schema.py:589]:
```suggestion
        def _is_1d_array(value):
            return isinstance(value, (list, np.ndarray)) and np.array(value).ndim == 1

```

File: mlflow/types/schema.py
Comment (serena-ruan) [reply mlflow/types/schema.py:579]: I think tuple works fine
image

File: mlflow/types/schema.py
Comment (serena-ruan) [reply mlflow/types/schema.py:472]: TensorSpec and ColSpec uses type, do we want to keep consistent?

File: mlflow/types/schema.py
Comment (serena-ruan) [reply mlflow/types/schema.py]: I think np.datetime64 is compatible with different precisions
image

File: mlflow/types/schema.py
Comment (BenWilson2) [inline mlflow/types/schema.py:711]: Can we add a test that validates this behavior? I think it's the only error handling that's not covered in the `test_pyfunc_schema_enforcement` suite below

File: mlflow/types/schema.py
Comment (BenWilson2) [reply mlflow/types/schema.py:472]: I think those are in need of a refactoring later on. ;)

File: mlflow/types/schema.py
Comment (BenWilson2) [reply mlflow/types/schema.py]: SGTM! Thank you for the validation :D

File: mlflow/types/schema.py
Comment (serena-ruan) [reply mlflow/types/schema.py:711]: Ok!

File: mlflow/types/schema.py
Comment (BenWilson2) [reply mlflow/types/schema.py:579]: It converts to a list :) 

``` python
import json

# Original tuple
original_tuple = (1, 2, 3)

# Serialize the tuple to a JSON string
serialized = json.dumps(original_tuple)

print(f'Serialized: {serialized}')

# Deserialize the JSON string back to a Python object
deserialized = json.loads(serialized)

print(f'Deserialized: {deserialized}')

# Check if the deserialized object equals the original tuple
print(f'Equality check: {original_tuple == deserialized}')
```

```sh
Serialized: [1, 2, 3]
Deserialized: [1, 2, 3]
Equality check: False
```

Comment (serena-ruan) [reply mlflow/types/schema.py:574]: Yes, but we convert it back to tuple when constructing the shape
https://github.com/mlflow/mlflow/blob/578d326f5af7ecfed412c2ee802e363d0d374c56/mlflow/types/schema.py#L477

Or you prefer we convert this to a list instead? I'm totally fine though :D

File: mlflow/types/schema.py
Comment (serena-ruan) [reply mlflow/types/schema.py]: Conclusion -> we always convert to python native types.

File: mlflow/types/schema.py
Comment (harupy) [inline mlflow/types/schema.py:511]:
```suggestion
    def enforce_param_datatype(cls, name, value, dtype: DataType):
```

File: mlflow/types/schema.py
Comment (harupy) [inline mlflow/types/schema.py:2]:
```suggestion
import datetime  # pylint: disable=unused-import
```

Let's remove this

**Review by harupy — APPROVED** [review]
looks good to me :)

File: mlflow/types/schema.py
Comment (serena-ruan) [reply mlflow/types/schema.py:2]: I tried 😢 I don't know why pylint keeps giving me this error
image

Comment (harupy) [reply mlflow/types/schema.py:2]: because it's unsed

Comment (serena-ruan) [reply mlflow/types/schema.py:2]: I used it for DataType.datetime's python type.
image

Comment (harupy) [reply mlflow/types/schema.py]: Maybe because it's shadowed by `datetime`?

Comment (serena-ruan) [reply mlflow/types/schema.py]: Updated to `from datetime import date`

Comment (harupy) [reply mlflow/types/schema.py]: sounds good

**Pull request merged**
